### DISCUSSION TITLE
Perceptron train

### DISCUSSION THREAD

**Discussion by lordpba** [Lightning-AI/dl-fundamentals#24] [Lecture videos and quizzes] [answered]
Hi!

I have done the Perceptron as in 1.6

but when I do the training I have the error: TypeError: can't multiply sequence by non-int of type 'float'

this is my code:

class Perceptron:
    def __init__(self, num_features):
        self.num_features = num_features
        self.weights = [0.0 for _ in range(num_features)]
        self.bias = 0.0
    
    def forward(self, x):
        weighted_sum_z = self.bias
        for i, _ in enumerate(self.weights):
            weighted_sum_z += x[i] * self.weights[i] #qui vengono moltiplicate tutte le features (x) per i corrispondenti pesi (w)
        
        if weighted_sum_z > 0.0: #qui applichiamo il 'treshold' dell'attivazione 0 o 1 se la somma dei pesi e' appunto 0 o 1
            prediction = 1
        else:
            prediction = 0
        
        return prediction
    
    def update(self, x, true_y):
        prediction = self.forward(x)
        error = true_y - prediction
        # da qui abbiamo l'aggiornamento dei pesi del perceptron dati dall'errore x la feature
        self.bias += error
        for i, _ in enumerate(self.weights):
            self.weights[i] += error * x[i]

        return error

def train(model, all_x, all_y, epochs):

    for epoch in range(epochs):
        error_count = 0

        for x, y in zip(all_x, all_y):
            error = model.update(x, y)
            error_count += abs(error)
        
        print(f"Epoch {epoch+1} errors {error_count}")

---------
ppn = Perceptron(num_features=2)

train(model=ppn, all_x=X_train, all_y=y_train, epochs=5)
----------

TypeError                                 Traceback (most recent call last)
Cell In[13], line 3
      1 ppn = Perceptron(num_features=2)
----> 3 train(model=ppn, all_x=X_train, all_y=y_train, epochs=5)

Cell In[11], line 7, in train(model, all_x, all_y, epochs)
      4 error_count = 0
      6 for x, y in zip(all_x, all_y):
----> 7     error = model.update(x, y)
      8     error_count += abs(error)
     10 print(f"Epoch {epoch+1} errors {error_count}")

Cell In[9], line 20, in Perceptron.update(self, x, true_y)
     19 def update(self, x, true_y):
---> 20     prediction = self.forward(x)
     21     error = true_y - prediction
     22     # da qui abbiamo l'aggiornamento dei pesi del perceptron dati dall'errore x la feature

Cell In[9], line 10, in Perceptron.forward(self, x)
      8 weighted_sum_z = self.bias
      9 for i, _ in enumerate(self.weights):
---> 10     weighted_sum_z += x[i] * self.weights[i] #qui vengono moltiplicate tutte le features (x) per i corrispondenti pesi (w)
     12 if weighted_sum_z > 0.0: #qui applichiamo il 'treshold' dell'attivazione 0 o 1 se la somma dei pesi e' appunto 0 o 1
     13     prediction = 1

TypeError: can't multiply sequence by non-int of type 'float

**Comment by rasbt** [comment, marked as answer]
Hmm, to help finding out what's going on, can you modify this

```python
    for i, _ in enumerate(self.weights):
        weighted_sum_z += x[i] * self.weights[i] 
```

to

```python
    for i, _ in enumerate(self.weights):
        print(x[i])
        weighted_sum_z += x[i] * self.weights[i] 
```

to find out what could be causing this?

**Reply by lordpba** [reply]
got it! it was an error by me in loading the dataset:

before:

X_train = df[['x1','x2']]
y_train = df['label']

after (correct):

X_train = df[["x1", "x2"]].values
y_train = df["label"].values

now all ok!

Thanks for quick reply!!!

**Reply by rasbt** [reply]
Nice, glad you could fix it!
